File: .github/skills/angular-design-system-analyzer/scripts/detect_patterns.py
#!/usr/bin/env python3
"""
Detect code patterns and conventions in an Angular design system.
Analyzes coding standards, architectural patterns, and best practices.
"""

import os
import json
import sys
import re
from pathlib import Path
from typing import Dict, List, Any
from collections import Counter


def detect_patterns(repo_path: str) -> Dict[str, Any]:
    """
    Detect patterns and conventions in the Angular project.

    Args:
        repo_path: Path to the Angular repository

    Returns:
        Dictionary containing pattern analysis
    """
    repo_path = Path(repo_path).resolve()

    if not repo_path.exists():
        raise ValueError(f"Repository path does not exist: {repo_path}")

    analysis = {
        "naming_conventions": {},
        "folder_structure_patterns": {},
        "import_patterns": {},
        "architectural_patterns": {},
        "code_standards": {},
        "common_decorators": {},
        "rxjs_patterns": [],
    }

    src_path = repo_path / "src"
    if not src_path.exists():
        raise ValueError(f"src directory not found in {repo_path}")

    # Collect all TypeScript files
    ts_files = list(src_path.rglob("*.ts"))

    # Analyze naming conventions
    analysis["naming_conventions"] = analyze_naming_conventions(ts_files, repo_path)

    # Analyze folder structure
    analysis["folder_structure_patterns"] = analyze_folder_structure(
        src_path, repo_path
    )

    # Analyze imports
    analysis["import_patterns"] = analyze_import_patterns(ts_files)

    # Detect architectural patterns
    analysis["architectural_patterns"] = detect_architectural_patterns(
        ts_files, repo_path
    )

    # Analyze code standards
    analysis["code_standards"] = analyze_code_standards(ts_files)

    # Detect common decorators
    analysis["common_decorators"] = detect_common_decorators(ts_files)

    # Detect RxJS patterns
    analysis["rxjs_patterns"] = detect_rxjs_patterns(ts_files)

    return analysis


def analyze_naming_conventions(ts_files: List[Path], repo_root: Path) -> Dict[str, Any]:
    """Analyze naming conventions used in the project."""
    conventions = {
        "file_naming": {},
        "class_naming": {},
        "variable_naming": {},
        "examples": {},
    }

    file_patterns = Counter()
    class_patterns = Counter()

    for ts_file in ts_files[:100]:  # Sample first 100 files
        # Analyze file naming
        file_name = ts_file.stem
        if "." in file_name:
            parts = file_name.split(".")
            if len(parts) >= 2:
                suffix = parts[-1]
                file_patterns[suffix] += 1

        # Analyze class naming
        try:
            content = ts_file.read_text(encoding="utf-8")
            class_matches = re.findall(r"export\s+class\s+(\w+)", content)
            for class_name in class_matches:
                # Detect naming pattern (PascalCase, camelCase, etc.)
                if class_name[0].isupper():
                    class_patterns["PascalCase"] += 1
                else:
                    class_patterns["camelCase"] += 1
        except:
            continue

    conventions["file_naming"] = {
        "pattern": "kebab-case with type suffix",
        "common_suffixes": dict(file_patterns.most_common(10)),
    }

    conventions["class_naming"] = {
        "pattern": (
            "PascalCase"
            if class_patterns.get("PascalCase", 0) > class_patterns.get("camelCase", 0)
            else "camelCase"
        ),
        "distribution": dict(class_patterns),
    }

    conventions["variable_naming"] = {
        "pattern": "camelCase (Angular standard)",
        "constants": "UPPER_SNAKE_CASE",
        "private_members": "prefixed with underscore or camelCase",
    }

    return conventions


def analyze_folder_structure(src_path: Path, repo_root: Path) -> Dict[str, Any]:
    """Analyze folder organization patterns."""
    structure = {
        "organization_type": "unknown",
        "common_folders": [],
        "depth_analysis": {},
        "module_organization": "unknown",
    }

    # Find common top-level folders in src
    if src_path.exists():
        folders = [
            d.name
            for d in src_path.iterdir()
            if d.is_dir() and not d.name.startswith(".")
        ]
        structure["common_folders"] = folders

        # Detect organization type
        if "app" in folders:
            app_path = src_path / "app"
            app_folders = [d.name for d in app_path.iterdir() if d.is_dir()]

            # Feature-based if has multiple feature folders
            feature_indicators = ["features", "modules", "pages", "views"]
            if any(indicator in app_folders for indicator in feature_indicators):
                structure["organization_type"] = "feature-based"
            # Shared/Core pattern
            elif "shared" in app_folders or "core" in app_folders:
                structure["organization_type"] = "shared-core pattern"
            else:
                structure["organization_type"] = "component-based"

    return structure


def analyze_import_patterns(ts_files: List[Path]) -> Dict[str, Any]:
    """Analyze import statement patterns."""
    patterns = {
        "relative_imports": 0,
        "absolute_imports": 0,
        "barrel_imports": 0,
        "most_imported_modules": Counter(),
        "import_aliases": {},
    }

    for ts_file in ts_files[:100]:  # Sample
        try:
            content = ts_file.read_text(encoding="utf-8")

            # Find all imports
            import_matches = re.findall(
                r'import\s+.*\s+from\s+[\'"]([^\'"]+)[\'"]', content
            )

            for import_path in import_matches:
                if import_path.startswith("."):
                    patterns["relative_imports"] += 1
                elif import_path.startswith("@"):
                    patterns["absolute_imports"] += 1
                    # Extract module name
                    module = import_path.split("/")[0]
                    patterns["most_imported_modules"][module] += 1
                else:
                    patterns["most_imported_modules"][import_path.split("/")[0]] += 1

                # Detect barrel imports (index files)
                if import_path.endswith("/index") or import_path.endswith(""):
                    patterns["barrel_imports"] += 1
        except:
            continue

    patterns["most_imported_modules"] = dict(
        patterns["most_imported_modules"].most_common(10)
    )

    return patterns


def detect_architectural_patterns(
    ts_files: List[Path], repo_root: Path
) -> Dict[str, Any]:
    """Detect architectural patterns used."""
    patterns = {
        "uses_services": False,
        "uses_state_management": False,
        "state_management_library": None,
        "uses_lazy_loading": False,
        "uses_standalone_components": False,
        "uses_dependency_injection": False,
        "uses_reactive_forms": False,
        "uses_template_driven_forms": False,
    }

    all_content = ""
    for ts_file in ts_files[:50]:  # Sample
        try:
            all_content += ts_file.read_text(encoding="utf-8")
        except:
            continue

    # Detect patterns
    if ".service" in all_content or "Injectable" in all_content:
        patterns["uses_services"] = True
        patterns["uses_dependency_injection"] = True

    if "ngrx" in all_content.lower() or "@ngrx" in all_content:
        patterns["uses_state_management"] = True
        patterns["state_management_library"] = "NgRx"
    elif "akita" in all_content.lower():
        patterns["uses_state_management"] = True
        patterns["state_management_library"] = "Akita"

    if "loadChildren" in all_content:
        patterns["uses_lazy_loading"] = True

    if "standalone: true" in all_content:
        patterns["uses_standalone_components"] = True

    if "FormGroup" in all_content or "FormControl" in all_content:
        patterns["uses_reactive_forms"] = True

    if "ngModel" in all_content:
        patterns["uses_template_driven_forms"] = True

    return patterns


def analyze_code_standards(ts_files: List[Path]) -> Dict[str, Any]:
    """Analyze code quality and standards."""
    standards = {
        "uses_typescript_strict": False,
        "uses_linting": False,
        "uses_prettier": False,
        "average_file_length": 0,
        "uses_jsdoc": False,
        "uses_interfaces": False,
        "uses_enums": False,
        "uses_type_aliases": False,
    }

    total_lines = 0
    file_count = 0

    for ts_file in ts_files[:50]:  # Sample
        try:
            content = ts_file.read_text(encoding="utf-8")
            lines = content.split("\n")
            total_lines += len(lines)
            file_count += 1

            # Detect standards
            if "interface " in content:
                standards["uses_interfaces"] = True
            if "enum " in content:
                standards["uses_enums"] = True
            if "type " in content and "=" in content:
                standards["uses_type_aliases"] = True
            if "/**" in content or "/*" in content:
                standards["uses_jsdoc"] = True
        except:
            continue

    if file_count > 0:
        standards["average_file_length"] = total_lines // file_count

    return standards


def detect_common_decorators(ts_files: List[Path]) -> Dict[str, int]:
    """Detect commonly used decorators."""
    decorators = Counter()

    decorator_pattern = r"@(\w+)\("

    for ts_file in ts_files[:100]:  # Sample
        try:
            content = ts_file.read_text(encoding="utf-8")
            matches = re.findall(decorator_pattern, content)
            decorators.update(matches)
        except:
            continue

    return dict(decorators.most_common(15))


def detect_rxjs_patterns(ts_files: List[Path]) -> List[Dict[str, Any]]:
    """Detect RxJS usage patterns."""
    patterns = []

    rxjs_operators = Counter()

    for ts_file in ts_files[:50]:  # Sample
        try:
            content = ts_file.read_text(encoding="utf-8")

            # Detect RxJS imports
            rxjs_import_match = re.search(
                r'import\s+{([^}]+)}\s+from\s+[\'"]rxjs', content
            )
            if rxjs_import_match:
                operators = [op.strip() for op in rxjs_import_match.group(1).split(",")]
                rxjs_operators.update(operators)
        except:
            continue

    if rxjs_operators:
        patterns.append(
            {
                "pattern": "RxJS Observables",
                "common_operators": dict(rxjs_operators.most_common(10)),
            }
        )

    return patterns


def main():
    if len(sys.argv) < 2:
        print("Usage: python detect_patterns.py <repo_path> [output_file]")
        print(
            "Example: python detect_patterns.py /path/to/angular-project patterns.json"
        )
        sys.exit(1)

    repo_path = sys.argv[1]
    output_file = sys.argv[2] if len(sys.argv) > 2 else "patterns_analysis.json"

    try:
        print(f"Detecting patterns in: {repo_path}")
        analysis = detect_patterns(repo_path)

        # Write to output file
        with open(output_file, "w", encoding="utf-8") as f:
            json.dump(analysis, f, indent=2, ensure_ascii=False)

        print(f"\n✅ Analysis complete! Results saved to: {output_file}")
        print(f"\nKey Findings:")
        print(
            f"  Organization: {analysis['folder_structure_patterns'].get('organization_type', 'unknown')}"
        )
        print(
            f"  Uses Services: {analysis['architectural_patterns'].get('uses_services', False)}"
        )
        print(
            f"  State Management: {analysis['architectural_patterns'].get('state_management_library', 'None')}"
        )

    except Exception as e:
        print(f"❌ Error: {e}", file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    main()

File: .agent/skills/angular-design-system-analyzer-v2/scripts/detect_patterns.mjs
#!/usr/bin/env node
/**
 * Detect code patterns and conventions in an Angular design system.
 */
import fs from 'node:fs';
import path from 'node:path';

function rglob(dir, ext) {
  const results = [];
  const skip = new Set(['node_modules', 'dist', 'coverage', '.git', '.angular']);
  function walk(d) {
    let entries;
    try { entries = fs.readdirSync(d, { withFileTypes: true }); } catch { return; }
    for (const e of entries) {
      if (e.name.startsWith('.') && e.name !== '.') continue;
      if (skip.has(e.name) && e.isDirectory()) continue;
      const fp = path.join(d, e.name);
      if (e.isDirectory()) walk(fp);
      else if (e.isFile() && e.name.endsWith(ext)) results.push(fp);
    }
  }
  walk(dir);
  return results;
}

function inc(o, k) { o[k] = (o[k] || 0) + 1; }
function topN(o, n) {
  return Object.fromEntries(Object.entries(o).sort((a, b) => b[1] - a[1]).slice(0, n));
}

function detectPatterns(repoPath) {
  const rp = path.resolve(repoPath);
  if (!fs.existsSync(rp)) throw new Error(`Path does not exist: ${rp}`);
  const srcPath = path.join(rp, 'src');
  if (!fs.existsSync(srcPath)) throw new Error(`src directory not found in ${rp}`);
  const tsFiles = rglob(srcPath, '.ts');

  return {
    naming_conventions: analyzeNaming(tsFiles),
    folder_structure_patterns: analyzeFolders(srcPath),
    import_patterns: analyzeImports(tsFiles),
    architectural_patterns: detectArch(tsFiles),
    code_standards: analyzeStandards(tsFiles),
    common_decorators: detectDecorators(tsFiles),
    rxjs_patterns: detectRxjs(tsFiles),
  };
}

function analyzeNaming(tsFiles) {
  const fp = {}, cp = {};
  for (const f of tsFiles.slice(0, 100)) {
    const fn = path.basename(f, '.ts');
    if (fn.includes('.')) { const s = fn.split('.'); inc(fp, s[s.length - 1]); }
    try {
      const c = fs.readFileSync(f, 'utf-8');
      for (const m of c.matchAll(/export\s+class\s+(\w+)/g)) {
        inc(cp, m[1][0] === m[1][0].toUpperCase() ? 'PascalCase' : 'camelCase');
      }
    } catch { continue; }
  }
  return {
    file_naming: { pattern: 'kebab-case with type suffix', common_suffixes: topN(fp, 10) },
    class_naming: { pattern: (cp.PascalCase || 0) > (cp.camelCase || 0) ? 'PascalCase' : 'camelCase', distribution: cp },
    variable_naming: { pattern: 'camelCase (Angular standard)', constants: 'UPPER_SNAKE_CASE', private_members: 'prefixed with underscore or camelCase' },
    examples: {},
  };
}

function analyzeFolders(srcPath) {
  const s = { organization_type: 'unknown', common_folders: [], depth_analysis: {}, module_organization: 'unknown' };
  if (!fs.existsSync(srcPath)) return s;
  const folders = fs.readdirSync(srcPath, { withFileTypes: true }).filter(d => d.isDirectory() && !d.name.startsWith('.')).map(d => d.name);
  s.common_folders = folders;
  if (folders.includes('app')) {
    const af = fs.readdirSync(path.join(srcPath, 'app'), { withFileTypes: true }).filter(d => d.isDirectory()).map(d => d.name);
    if (['features', 'modules', 'pages', 'views'].some(i => af.includes(i))) s.organization_type = 'feature-based';
    else if (af.includes('shared') || af.includes('core')) s.organization_type = 'shared-core pattern';
    else s.organization_type = 'component-based';
  }
  return s;
}

function analyzeImports(tsFiles) {
  const p = { relative_imports: 0, absolute_imports: 0, barrel_imports: 0, most_imported_modules: {}, import_aliases: {} };
  for (const f of tsFiles.slice(0, 100)) {
    let c; try { c = fs.readFileSync(f, 'utf-8'); } catch { continue; }
    for (const m of c.matchAll(/import\s+.*\s+from\s+['"]([^'"]+)['"]/g)) {
      const ip = m[1];
      if (ip.startsWith('.')) p.relative_imports++;
      else if (ip.startsWith('@')) { p.absolute_imports++; inc(p.most_imported_modules, ip.split('/')[0]); }
      else inc(p.most_imported_modules, ip.split('/')[0]);
      if (ip.endsWith('/index')) p.barrel_imports++;
    }
  }
  p.most_imported_modules = topN(p.most_imported_modules, 10);
  return p;
}

function detectArch(tsFiles) {
  const p = { uses_services: false, uses_state_management: false, state_management_library: null, uses_lazy_loading: false, uses_standalone_components: false, uses_dependency_injection: false, uses_reactive_forms: false, uses_template_driven_forms: false };
  let all = '';
  for (const f of tsFiles.slice(0, 50)) { try { all += fs.readFileSync(f, 'utf-8'); } catch { continue; } }
  if (all.includes('.service') || all.includes('Injectable')) { p.uses_services = true; p.uses_dependency_injection = true; }
  if (all.toLowerCase().includes('ngrx') || all.includes('@ngrx')) { p.uses_state_management = true; p.state_management_library = 'NgRx'; }
  else if (all.toLowerCase().includes('akita')) { p.uses_state_management = true; p.state_management_library = 'Akita'; }
  if (all.includes('loadChildren')) p.uses_lazy_loading = true;
  if (all.includes('standalone: true')) p.uses_standalone_components = true;
  if (all.includes('FormGroup') || all.includes('FormControl')) p.uses_reactive_forms = true;
  if (all.includes('ngModel')) p.uses_template_driven_forms = true;
  return p;
}

function analyzeStandards(tsFiles) {
  const s = { uses_typescript_strict: false, uses_linting: false, uses_prettier: false, average_file_length: 0, uses_jsdoc: false, uses_interfaces: false, uses_enums: false, uses_type_aliases: false };
  let tl = 0, fc = 0;
  for (const f of tsFiles.slice(0, 50)) {
    let c; try { c = fs.readFileSync(f, 'utf-8'); } catch { continue; }
    tl += c.split('\n').length; fc++;
    if (c.includes('interface ')) s.uses_interfaces = true;
    if (c.includes('enum ')) s.uses_enums = true;
    if (c.includes('type ') && c.includes('=')) s.uses_type_aliases = true;
    if (c.includes('/**') || c.includes('/*')) s.uses_jsdoc = true;
  }
  if (fc > 0) s.average_file_length = Math.floor(tl / fc);
  return s;
}

function detectDecorators(tsFiles) {
  const d = {};
  for (const f of tsFiles.slice(0, 100)) {
    let c; try { c = fs.readFileSync(f, 'utf-8'); } catch { continue; }
    for (const m of c.matchAll(/@(\w+)\(/g)) inc(d, m[1]);
  }
  return topN(d, 15);
}

function detectRxjs(tsFiles) {
  const ops = {};
  for (const f of tsFiles.slice(0, 50)) {
    let c; try { c = fs.readFileSync(f, 'utf-8'); } catch { continue; }
    const m = c.match(/import\s+{([^}]+)}\s+from\s+['"]rxjs/);
    if (m) for (const op of m[1].split(',').map(s => s.trim())) inc(ops, op);
  }
  if (Object.keys(ops).length > 0) return [{ pattern: 'RxJS Observables', common_operators: topN(ops, 10) }];
  return [];
}

// --- Main ---
const args = process.argv.slice(2);
if (args.length < 1) { console.log('Usage: node detect_patterns.mjs <repo_path> [output_file]'); process.exit(1); }
const repoPath = args[0], outputFile = args[1] || 'patterns_analysis.json';
try {
  console.log(`Detecting patterns in: ${repoPath}`);
  const analysis = detectPatterns(repoPath);
  fs.writeFileSync(outputFile, JSON.stringify(analysis, null, 2), 'utf-8');
  console.log(`\n✅ Analysis complete! Results saved to: ${outputFile}`);
  console.log(`\nKey Findings:`);
  console.log(`  Organization: ${analysis.folder_structure_patterns.organization_type || 'unknown'}`);
  console.log(`  Uses Services: ${analysis.architectural_patterns.uses_services}`);
  console.log(`  State Management: ${analysis.architectural_patterns.state_management_library || 'None'}`);
} catch (e) { console.error(`❌ Error: ${e.message}`); process.exit(1); }
